### PR TITLE
grab bag of small UB fixes

### PR DESCRIPTION
**Pull request by spoonincode**
Brings in,
AntelopeIO/abieos#25
AntelopeIO/appbase#31
AntelopeIO/bn256#20
AntelopeIO/eos-vm#22
along with some other noted changes below.

### CODE REVIEW

File: libraries/chain/include/eosio/chain/webassembly/eos-vm-oc/config.hpp [hunk 1]
```diff
@@ -20,7 +24,11 @@ struct config {
    // libtester disables the limits in all tests, except enforces the limits
    // in the tests in unittests/eosvmoc_limits_tests.cpp.
    std::optional<rlim_t>   cpu_limit {20u};
+#if __has_feature(undefined_behavior_sanitizer) || __has_feature(address_sanitizer)
+   std::optional<rlim_t>   vm_limit; // UBSAN & ASAN can add massive virtual memory usage; don't enforce vm limits when either of them are enabled
```
Comment (spoonincode) [inline libraries/chain/include/eosio/chain/webassembly/eos-vm-oc/config.hpp:28]: Don't need this for tests to pass: OC in unit tests always has `vm_limit` disabled now (except for the one modified test case), and python integration tests never use OC. But I'd like to use OC + ASAN/UBSAN with nodeos. Such usage is probably "developer enough" where simply disabling this limit is safe when configured that way; i.e. won't trap unsuspecting users. Alternatively could guard this disablement behind `EOSVMOC_ENABLE_DEVELOPER_OPTIONS` but ultimately I want to remove that option.

File: libraries/chain/webassembly/runtimes/eos-vm-oc/memory.cpp [hunk 2]
```diff
@@ -40,7 +40,7 @@ memory::memory(uint64_t sliced_pages) {
    uintptr_t* const intrinsic_jump_table = reinterpret_cast<uintptr_t* const>(zeropage_base - first_intrinsic_offset);
    const intrinsic_map_t& intrinsics = get_intrinsic_map();
    for(const auto& intrinsic : intrinsics)
-      intrinsic_jump_table[-intrinsic.second.ordinal] = (uintptr_t)intrinsic.second.function_ptr;
+      intrinsic_jump_table[-(int)intrinsic.second.ordinal] = (uintptr_t)intrinsic.second.function_ptr;
```
Comment (spoonincode) [inline libraries/chain/webassembly/runtimes/eos-vm-oc/memory.cpp:43]: This is an interesting one. The error here gets reported somewhat confusingly as `addition of unsigned offset to 0x7aca262015f8 overflowed to 0x7aca26201570`. I believe what is occurring is that `ordinal` is a `size_t` thus `uint64_t`, and negating it is still a `uint64_t`: now a really large `uint64_t`. But pointer arithmetic is defined on integer math; `ptrdiff_t` is a signed integer for example. So I think this is causing some sort of wrap around on a large signed integer. Casting `ordinal` to an `int` before the negation ensures it remains a small value `int` before pointer arithmetic. I'd be curious on other takes though.

**Comment by spoonincode** [comment]
whoops, I pushed b47a64c here to the wrong branch; I wanted to have a separate PR for that. Sorry for force pushing it away...

File: libraries/chain/webassembly/runtimes/eos-vm-oc/memory.cpp (hunk 2 above)
Comment (spoonincode) [reply libraries/chain/webassembly/runtimes/eos-vm-oc/memory.cpp:43]: `ordinal` is the host function index over in,
https://github.com/AntelopeIO/leap/blob/968a0fa0d0aa5c36ea2f2ed3eb4469068a9afd68/libraries/chain/include/eosio/chain/webassembly/eos-vm-oc/intrinsic_mapping.hpp#L18-L19
so `int` will always be plenty large enough (we won't have 2<sup>31</sup> host functions). My initial impression of `intptr_t` is that its purpose/connotation isn't really for indexes in to an array.

I do wonder if `ordinal` should just be an `int` instead; I don't think it'd be too hard to change it.

**Comment by ericpassmore** [comment]
Note:start
group: CLEANCODE
category: INTERNALS
summary: Fixes several undefined behavior (UB) bugs. 
Note: end

**Review by heifner — APPROVED** [review]

**Review by greg7mdp — APPROVED** [review]

**Pull request merged**